haystack/nodes/extractor/entity.py

```diff
@@ -38,90 +67,393 @@ class EntityExtractor(BaseComponent):
+        self.aggregation_strategy = aggregation_strategy
+        self.ignore_labels = ignore_labels
+
+        if add_prefix_space is None:
+            tokenizer = AutoTokenizer.from_pretrained(model_name_or_path, use_auth_token=use_auth_token)
+        else:
+            tokenizer = AutoTokenizer.from_pretrained(
+                model_name_or_path, use_auth_token=use_auth_token, add_prefix_space=add_prefix_space
+            )
+        if not tokenizer.is_fast:
+            logger.error(
+                "The EntityExtractor node only works when using a fast tokenizer. Please choose a model "
+                "that has a corresponding fast tokenizer."
+            )
+
+        self.tokenizer = tokenizer
+        self.max_seq_len = max_seq_len if max_seq_len else self.tokenizer.model_max_length
+
+        self.pre_split_text = pre_split_text
+        self.pre_tokenizer = WhitespaceSplit()
+
+        self.model = AutoModelForTokenClassification.from_pretrained(
+            model_name_or_path, use_auth_token=use_auth_token, revision=model_version
+        )
+        self.model.to(str(self.devices[0]))
+        self.entity_postprocessor = _EntityPostProcessor(model=self.model, tokenizer=self.tokenizer)
+
+    @staticmethod
+    def _add_entities_to_doc(
+        doc: Union[Document, dict], entities: List[dict], flatten_entities_in_meta_data: bool = False
+    ):
+        """Add the entities to the metadata of the document.
+
+        :param doc: The document where the metadata will be added.
+        :param entities: The list of entities predicted for document `doc`.
+        :param flatten_entities_in_meta_data: If True this converts all entities predicted for a document from a list of
+            dictionaries into a single list for each key in the dictionary.
+        """
+        is_doc = isinstance(doc, Document)
+        if flatten_entities_in_meta_data:
+            new_key_map = {
+                "entity_group": "entity_groups",
+                "score": "entity_scores",
+                "word": "entity_words",
+                "start": "entity_starts",
+                "end": "entity_ends",
+            }
+            entity_lists: Dict[str, List[Any]] = {v: [] for k, v in new_key_map.items()}
+            for entity in entities:
+                for key in entity:
+                    new_key = new_key_map[key]
+                    if isinstance(entity[key], np.float32):
+                        entity_lists[new_key].append(float(entity[key]))
+                    else:
+                        entity_lists[new_key].append(entity[key])
+            if is_doc:
+                doc.meta.update(entity_lists)  # type: ignore
+            else:
+                doc["meta"].update(entity_lists)  # type: ignore
+        else:
+            if is_doc:
+                doc.meta["entities"] = entities  # type: ignore
+            else:
+                doc["meta"]["entities"] = entities  # type: ignore
```

Comment:
Why do we do this? It seems somehow "unnatural" to me :-) If I pass documents to the entity extractor then I would expect to get results back. As we currently do. But to get results and change the meta of the documents as well, at the same time, is somehow unexpected to me. We should at least provide a kwarg in the run method or a boolean switch in the init, wdyt?

Reply:
This was the behavior of the node before. And for this node to work in indexing pipelines the node needs to return documents, which is why we decided to add the predictions into the meta data.

Reply:
Do you think it is fine that we leave the `run` and `run_batch` methods to always return documents with updated meta data. And then if the user just wants the results they can choose to run the `extract` method directly. What do you think?

Reply:
Conclusion: we change the Document meta field in several components, and sometimes we return new objects. Sometimes we mix (change the meta field and then return the documents). Let's go with things as is; I noted this on our to-do list.